core/src/main/java/ai/timefold/solver/core/impl/score/director/ValueRangeManager.java

```diff
@@ -358,59 +365,255 @@ public <T> CountableValueRange<T> getFromSolution(ValueRangeDescriptor<Solution_
+            throw new IllegalStateException(
+                    "Impossible state: the value range (%s) with sorter (%s) does not align with the existing ascending (%s) and descending (%s) sorters."
+                            .formatted(valueRangeDescriptor, sorter, leftSorter, rightSorter));
+        }
+    }
+
+    private Map<Object, Integer> getIndexMapFromSolution(ValueRangeDescriptor<Solution_> valueRangeDescriptor) {
+        getFromSolution(valueRangeDescriptor);
+        return Objects.requireNonNull(fromSolutionValueIndexMap[valueRangeDescriptor.getOrdinal()]);
+    }
+
+    private <T> CountableValueRange<T> fetchValueRangeFromSolution(ValueRangeDescriptor<Solution_> valueRangeDescriptor,
+            Solution_ solution, @Nullable SelectionSorter<Solution_, T> sorter) {
+        CountableValueRange<T> valueRange;
+        var extractedValueRange = valueRangeDescriptor.<T> extractAllValues(Objects.requireNonNull(solution));
+        if (!(extractedValueRange instanceof CountableValueRange<T> countableValueRange)) {
+            throw new UnsupportedOperationException("""
+                    Impossible state: value range (%s) on planning solution (%s) is not countable.
+                    Maybe replace %s with %s."""
+                    .formatted(valueRangeDescriptor, solution, DoubleValueRange.class.getSimpleName(),
+                            BigDecimalValueRange.class.getSimpleName()));
+        } else if (valueRangeDescriptor.acceptsNullInValueRange()) {
+            valueRange = new NullAllowingCountableValueRange<>(countableValueRange);
+        } else {
+            valueRange = countableValueRange;
+        }
+        if (sorter != null) {
+            if (!(valueRange instanceof SortableValueRange sortableValueRange)) {
+                throw new IllegalStateException(
+                        "Impossible state: value range (%s) on planning solution (%s) is not sortable."
+                                .formatted(valueRangeDescriptor, solution));
             }
-            fromSolution[valueRangeDescriptor.getOrdinal()] = valueRange;
+            var sorterAdapter = SelectionSorterAdapter.of(solution, sorter);
+            valueRange = (CountableValueRange<T>) sortableValueRange.sort(sorterAdapter);
         }
-        return (CountableValueRange<T>) valueRange;
+        return valueRange;
     }
 
     /**
      * @throws IllegalStateException if called before {@link #reset(Object)} is called
      */
     @SuppressWarnings("unchecked")
     public <T> CountableValueRange<T> getFromEntity(ValueRangeDescriptor<Solution_> valueRangeDescriptor, Object entity) {
+        return getFromEntity(valueRangeDescriptor, entity, null);
+    }
+
+    /**
+     * @throws IllegalStateException if called before {@link #reset(Object)} is called
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public <T> CountableValueRange<T> getFromEntity(ValueRangeDescriptor<Solution_> valueRangeDescriptor, Object entity,
```

Comment:
This method is very long. Surely there are self-contained pieces that could be split off into other methods and called from here.